src/block_merging/mod.rs

```diff
@@ -0,0 +1,770 @@
+                    CancunPayloadFields { parent_beacon_block_root, versioned_hashes: blob_versioned_hashes },
+                    PraguePayloadFields {
+                        requests: RequestsOrHash::Requests(response.execution_requests.to_requests()),
+                    },
+                ),
+            })?;
+
+            self.validation.validate_message_against_block(block, message, 0, false, None).await?;
+        }
+
+        Ok(response)
+    }
+
+    /// Merge a block by appending mergeable orders.
+    /// Returns the response with the block, the versioned hashes of the appended blobs,
+    /// and the cached reads used during execution.
+    async fn merge_block(
+        &self,
+        original_value: U256,
+        proposer_fee_recipient: Address,
+        base_block: Block,
+        merging_data: Vec<MergeableOrderBytes>,
+    ) -> Result<(BlockMergeResponseV1, Vec<B256>, CachedReads), BlockMergingApiError> {
+        info!(target: "rpc::relay", "Merging block v1");
+        let validation = &self.validation;
+        let evm_config = &validation.evm_config;
+
+        // Recover the base block transactions in parallel
+        let (base_block, senders) =
+            base_block.try_into_recovered().map_err(|_| BlockMergingApiError::InvalidSignatureInBaseBlock)?.split();
+
+        let (header, body) = base_block.split();
+
+        let (withdrawals, transactions) = (body.withdrawals, body.transactions);
+
+        let block_base_fee_per_gas = header.base_fee_per_gas.unwrap_or_default();
+
+        let relay_fee_recipient = self.relay_fee_recipient;
+        let beneficiary = header.beneficiary;
+
+        // Check we have collateral for this builder
+        let Some(types::PrivateKeySigner(signer)) = self.builder_collateral_map.get(&beneficiary).as_ref() else {
+            return Err(BlockMergingApiError::NoSignerForBuilder(beneficiary));
+        };
+
+        // Check that block has proposer payment, otherwise reject it.
+        // We don't remove it from the block, but add another payment transaction at the end.
+        let Some(payment_tx) = transactions.last() else {
+            return Err(BlockMergingApiError::MissingProposerPayment);
+        };
+        if payment_tx.value() != original_value || payment_tx.to() != Some(proposer_fee_recipient) {
+            return Err(BlockMergingApiError::InvalidProposerPayment);
+        }
+
+        let payment_tx_gas_limit = payment_tx.gas_limit();
+
+        // Leave some gas for the final revenue distribution call
+        // and the proposer payment.
+        // The gas cost should be 10k per target, but could jump
+        // to 35k if the targets are new accounts.
+        // This number leaves us space for ~9 non-empty targets, or ~2 new accounts.
+        // TODO: compute dynamically by keeping track of gas cost
+        // We also leave some gas for the proposer payment
+        let distribution_gas_limit = 100000 + payment_tx_gas_limit;
```

Comment:
Is it OK to use the payment transaction gas limit here? The downside of this is that it might be a lot higher than what's used, and we're currently estimating the payment cost from the gas limit.